auth/authenticator.hh

```diff
@@ -120,7 +128,7 @@ public:
     ///
     virtual const resource_set& protected_resources() const = 0;
 
-    virtual ::shared_ptr<sasl_challenge> new_sasl_challenge() const = 0;
+    virtual ::shared_ptr<sasl_challenge> new_sasl_challenge(session_dn_func = {}) const = 0;
 };
```

Comment:
It's wrong to insert a TLS related function into the interface. IIUC SASL isn't tied to TLS in any way. 
As a result password_authenticator receives this session_dn_func which it doesn't know what to do with.

Can't we use it in the concrete class that implements the interface?

Reply:
As stated below, certificates != TLS. The interface could just as well be new_sasl_challenge(std::vector<std::string> dn), but an optional callback saves work when not needed. Point is still just augmenting the sasl call with an option to derive data from certificate info, not making it wire aware. What I could agree with is that we could separate the interfaces and instead of trying to call something called SASL direct from server.cc, have a separately named interface. But it seems a little overzealous also...